### ISSUE TITLE
Add support for constrained optimization

### ISSUE THREAD

**Issue by SimonBlanke** [SimonBlanke/Gradient-Free-Optimizers#33]
I want to add the support for constrained optimization. The interface design could look like this:

```python
import numpy as np
from gradient_free_optimizers import HillClimbingOptimizer

def parabola_function(para):
    loss = para["x1"] * para["x1"] + para["x2"] * para["x2"]
    return -loss

search_space = {
    "x1": np.arange(-10, 10, 0.1),
    "x2": np.arange(-10, 10, 0.1),
}

def constraint_1(para):
    return para["x1"] ** 2 + para["x2"] ** 2 <= 2

constraints_list = [constraint_1]

opt = HillClimbingOptimizer(search_space, constraints=constraints_list)
opt.search(parabola_function, n_iter=100000)
```

Implementing this in the optimization algorithms will need various solutions:
 - For hill-climbing based algorithms I could just repeat the iteration step if the new position is not inside the constrained search-space
 - Sequence-mode-based optimizers might be the easiest to implement, because I could just remove the positions outside the constrained search-space from the list of all possible positions. Therefore the surrogate model would not make a prediction for those positions.
 - The difficult cases are the more deterministic algorithms like PSO or DIRECT. Repeating the iteration will just return the same position, so I'll have to think about this one.

**Comment by SimonBlanke** [comment]
Here is an example search-path-gif with a constraint at the optimum around the [0, 0] position for the Hill-climbing algorithm:

![HillClimbingOptimizer](https://user-images.githubusercontent.com/20640612/209572098-4c456417-745b-41e9-b46c-23effbec3743.gif)

**Comment by SimonBlanke** [comment]
A similar example for Bayesian Optimization. Very interesting to see:

![BayesianOptimizer](https://user-images.githubusercontent.com/20640612/209572236-38b0521d-483a-4b11-9d93-166b44274feb.gif)

**Comment by SimonBlanke** [comment]
A similar example for Grid Search. That way easier than expected:

![GridSearchOptimizer](https://user-images.githubusercontent.com/20640612/210131932-bfc8baf0-ff2c-4cce-ad9d-c22ae4a47bf0.gif)

**Comment by SimonBlanke** [comment]
Multiple constraints (x*x+y*y > 5  && x > -3) are not a problem:

![GridSearchOptimizer_constr](https://user-images.githubusercontent.com/20640612/229557998-932b3a4c-509d-48e6-8481-cc5788e47fa9.gif)

**Comment by SimonBlanke** [comment]
From my current progress I see the following optimization algorithms as non-trivial for the constraint optimization implementation:
- DownhillSimplexOptimizer
- PatternSearch
- EvolutionStrategyOptimizer
- DirectAlgorithm

**Comment by SimonBlanke** [comment]
I added an example for constrained optimization to the repository and released v1.3 for the [documentation](https://simonblanke.github.io/gradient-free-optimizers-documentation) for GFO. All tests look good.

I just released version 1.3 and will close this issue.

This is a big new feature, so feedback is appreciated and can be submitted in this issue.
